File: src/Solutions/Euler028.hs
module Solutions.Euler028 where
import Control.Monad
import Control.Monad.ST
import Data.STRef

compute = runST $ do
    sum <- newSTRef 1 
    step <- newSTRef 2
    curr <- newSTRef 1
    replicateM_ 500 $ do
        replicateM_ 4 $ do
            readSTRef step >>= modifySTRef curr . (+)
            readSTRef curr >>= modifySTRef sum . (+)
        modifySTRef step (+2)
    readSTRef sum

euler028 :: IO String
euler028 = return $ show compute

File: rust/src/solutions/euler028.rs
pub fn euler028() -> String {
    let mut sum = 1;
    let mut step = 2;
    let mut curr = 1;
    for _ in 0..500 {
        for _ in 0..4 {
            curr += step;
            sum += curr;
        }
        step += 2;
    }
    sum.to_string()
}
